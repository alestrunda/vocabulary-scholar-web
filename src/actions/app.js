export const APP_INFO_MESSAGE = 'APP_INFO_MESSAGE',
  APP_LOAD_STATE = 'APP_LOAD_STATE',
  APP_SET_ITEMS_PER_PAGE = 'APP_SET_ITEMS_PER_PAGE',
  APP_SET_LANGUAGE = 'APP_SET_LANGUAGE',
  APP_SET_SELECTED_PHRASE = 'APP_SET_SELECTED_PHRASE',
  APP_STATE_NOT_LOADED = 'APP_STATE_NOT_LOADED',
  APP_STORAGE_TYPE = 'APP_STORAGE_TYPE',
  APP_WARNING = 'APP_WARNING',
  APP_WARNING_CLOSE = 'APP_WARNING_CLOSE',
  APP_WORD_LIST_SET_PAGE = 'APP_WORD_LIST_SET_PAGE';

export const appInfoMessage = infoMessage => ({
  type: APP_INFO_MESSAGE,
  payload: { infoMessage },
});

export const appLoadState = (stateToLoad, storageType) => ({
  type: APP_LOAD_STATE,
  payload: { stateToLoad, storageType },
});

export const appSetItemsPerPage = number => ({
  type: APP_SET_ITEMS_PER_PAGE,
  payload: { number },
});

export const appSetLanguage = language => ({
  type: APP_SET_LANGUAGE,
  payload: { language },
});

export const appSetSelectedPhrase = phrase => ({
  type: APP_SET_SELECTED_PHRASE,
  payload: { phrase },
});

export const appStateNotLoaded = () => ({
  type: APP_STATE_NOT_LOADED,
});

export const appStorageType = storageType => ({
  type: APP_STORAGE_TYPE,
  payload: { storageType },
});

export const appWarning = (warning, warningType) => ({
  type: APP_WARNING,
  payload: { warning, warningType },
});

export const appWarningClose = warningType => ({
  type: APP_WARNING_CLOSE,
  payload: { warningType },
});

export const appWordListSetPage = (pageNumber, listID) => ({
  type: APP_WORD_LIST_SET_PAGE,
  payload: { pageNumber, listID },
});
