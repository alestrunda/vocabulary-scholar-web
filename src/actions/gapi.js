export const GAPI_CLIENT_STATUS_UPDATE = 'GAPI_CLIENT_STATUS_UPDATE',
  GAPI_CONNECTED = 'GAPI_CONNECTED',
  GAPI_SET_STORE_FILE_ID = 'GAPI_SET_STORE_FILE_ID',
  GAPI_STORE_LOADED_FROM_DRIVE = 'GAPI_STORE_LOADED_FROM_DRIVE';

export const gapiClientStatusUpdate = (isClientSignedIn, isInitial) => ({
  type: GAPI_CLIENT_STATUS_UPDATE,
  payload: { isClientSignedIn, isInitial },
});

export const gapiConnected = () => ({
  type: GAPI_CONNECTED,
});

export const gapiSetStoreFileID = storeFileID => ({
  type: GAPI_SET_STORE_FILE_ID,
  payload: { storeFileID },
});

export const gapiStoreLoadedFromDrive = (
  isStoreLoadedFromDrive,
  storeFileID
) => ({
  type: GAPI_STORE_LOADED_FROM_DRIVE,
  payload: { isStoreLoadedFromDrive, storeFileID },
});
