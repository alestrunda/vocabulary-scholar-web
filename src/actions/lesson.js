import uniqid from 'uniqid';

export const LESSON_CREATE = 'LESSON_CREATE',
  LESSON_ANONYMOUS_CREATE = 'LESSON_ANONYMOUS_CREATE',
  LESSON_LIST_REMOVE = 'LESSON_LIST_REMOVE',
  LESSON_REMOVE = 'LESSON_REMOVE',
  LESSON_RESET_SCORE = 'LESSON_RESET_SCORE',
  LESSON_SET_SCORE = 'LESSON_SET_SCORE',
  LESSON_VIEW = 'LESSON_VIEW',
  LESSONS_DELETE = 'LESSONS_DELETE',
  LESSONS_IMPORT = 'LESSONS_IMPORT';

export const lessonCreate = ({
  id = uniqid(),
  name = '',
  listIDs = [],
  questionsCnt = 0,
  dateCreated = new Date().getTime(),
  type,
  rateBy,
} = {}) => ({
  type: LESSON_CREATE,
  payload: { id, name, listIDs, questionsCnt, dateCreated, type, rateBy },
});

export const lessonAnonymousCreate = lesson => ({
  type: LESSON_ANONYMOUS_CREATE,
  payload: { lesson },
});

export const lessonListRemove = (lessonID, listID) => ({
  type: LESSON_LIST_REMOVE,
  payload: { lessonID, listID },
});

export const lessonRemove = lessonID => ({
  type: LESSON_REMOVE,
  payload: { lessonID },
});

export const lessonResetScore = lessonID => ({
  type: LESSON_RESET_SCORE,
  payload: { lessonID },
});

export const lessonSetScore = (lessonID, score, finishedIn) => ({
  type: LESSON_SET_SCORE,
  payload: { lessonID, score, finishedIn },
});

export const lessonView = (lessonID, dateViewed = new Date().getTime()) => ({
  type: LESSON_VIEW,
  payload: { lessonID, dateViewed },
});

export const lessonsDelete = () => ({
  type: LESSONS_DELETE,
});

export const lessonsImport = (lessons, dateCreated = new Date().getTime()) => ({
  type: LESSONS_IMPORT,
  payload: { lessons, dateCreated },
});
