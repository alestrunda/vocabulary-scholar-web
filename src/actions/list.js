import uniqid from 'uniqid';

export const LIST_CREATE = 'LIST_CREATE',
  LIST_REMOVE = 'LIST_REMOVE',
  LIST_SET_WORDS = 'LIST_SET_WORDS',
  LIST_SORT = 'LIST_SORT',
  LIST_SORT_RESET = 'LIST_SORT_RESET',
  LIST_VIEW = 'LIST_VIEW',
  LISTS_DELETE = 'LISTS_DELETE',
  LISTS_IMPORT = 'LISTS_IMPORT';

export const listCreate = ({
  id = uniqid(),
  name = '',
  wordIDs = [],
  dateCreated = new Date().getTime(),
} = {}) => ({
  type: LIST_CREATE,
  payload: { id, name, wordIDs, dateCreated },
});

export const listRemove = listID => ({
  type: LIST_REMOVE,
  payload: { listID },
});

export const listSetWords = (listID, selectedWordIDs) => ({
  type: LIST_SET_WORDS,
  payload: { listID, selectedWordIDs },
});

export const listSort = (
  listID,
  sortBy,
  isDescSort,
  words,
  forceResort = false
) => ({
  type: LIST_SORT,
  payload: { forceResort, listID, sortBy, isDescSort, words },
});

export const listSortReset = listID => ({
  type: LIST_SORT_RESET,
  payload: { listID },
});

export const listView = (listID, dateViewed = new Date().getTime()) => ({
  type: LIST_VIEW,
  payload: { listID, dateViewed },
});

export const listsDelete = () => ({
  type: LISTS_DELETE,
});

export const listsImport = (lists, dateCreated = new Date().getTime()) => ({
  type: LISTS_IMPORT,
  payload: { lists, dateCreated },
});
