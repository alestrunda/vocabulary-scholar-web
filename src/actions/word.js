import uniqid from 'uniqid';

export const WORD_ADD_PHRASE = 'WORD_ADD_PHRASE',
  WORD_EDIT_PHRASE = 'WORD_EDIT_PHRASE',
  WORD_LIST_SELECTED = 'WORD_LIST_SELECTED',
  WORD_REMOVE_PHRASE = 'WORD_REMOVE_PHRASE',
  WORD_SET_CUSTOM_FIELD = 'WORD_SET_CUSTOM_FIELD',
  WORD_SET_FAVORITE = 'WORD_SET_FAVORITE',
  WORD_SET_IMAGE_SRC = 'WORD_SET_IMAGE_SRC',
  WORD_SET_AUDIO_SRC = 'WORD_SET_AUDIO_SRC',
  WORD_SET_RATING = 'WORD_SET_RATING',
  WORD_VIEW = 'WORD_VIEW',
  WORDS_DELETE = 'WORDS_DELETE',
  WORDS_IMPORT = 'WORDS_IMPORT';

export const wordAddPhrase = ({
  id = uniqid(),
  wordID,
  phrase,
  startIndex,
  endIndex,
} = {}) => ({
  type: WORD_ADD_PHRASE,
  payload: { id, wordID, phrase, startIndex, endIndex },
});

export const wordEditPhrase = ({
  wordID,
  phraseID,
  phrase,
  startIndex,
  endIndex,
} = {}) => ({
  type: WORD_EDIT_PHRASE,
  payload: { wordID, phraseID, phrase, startIndex, endIndex },
});

export const wordListSelected = (wordID, listID) => ({
  type: WORD_LIST_SELECTED,
  payload: { wordID, listID },
});

export const wordRemovePhrase = (wordID, phraseID) => ({
  type: WORD_REMOVE_PHRASE,
  payload: { wordID, phraseID },
});

export const wordSetCustomField = (wordID, customField) => ({
  type: WORD_SET_CUSTOM_FIELD,
  payload: { wordID, customField },
});

export const wordSetFavorite = ({
  dateAdded = new Date().getTime(),
  wordID,
  word,
  isSettingAsFavorite = false,
} = {}) => ({
  type: WORD_SET_FAVORITE,
  payload: { wordID, dateAdded, word, isSettingAsFavorite },
});

export const wordSetImageSrc = (wordID, imageSrc) => ({
  type: WORD_SET_IMAGE_SRC,
  payload: { wordID, imageSrc },
});

export const wordSetAudioSrc = (wordID, audioSrc) => ({
  type: WORD_SET_AUDIO_SRC,
  payload: { wordID, audioSrc },
});

export const wordSetRating = ({
  wordID,
  rating,
  dateRated = new Date().getTime(),
} = {}) => ({
  type: WORD_SET_RATING,
  payload: { wordID, rating, dateRated },
});

export const wordView = (wordID, dateViewed = new Date().getTime()) => ({
  type: WORD_VIEW,
  payload: { wordID, dateViewed },
});

export const wordsDelete = () => ({
  type: WORDS_DELETE,
});

export const wordsImport = (words, dateAdded = new Date().getTime()) => ({
  type: WORDS_IMPORT,
  payload: { words, dateAdded },
});
