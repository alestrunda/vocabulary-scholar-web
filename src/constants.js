export const APP_VERSION = '1.0.0';

export const LIST_MAIN_ID = 'main';

export const STORAGE_FILENAME = 'vocabulary-scholar';

export const DATE_FORMAT = 'mmm d. yyyy';

export const WORD_RATINGS_CNT = 5;
export const DEFAULT_LESSON_QUESTIONS_CNT = 30;
export const MIN_WORD_LENGTH_SEARCH = 2;

export const SNACKBAR_HIDE_DURATION = 4000; //ms

export const WEEK_IN_MS = 604800000; //1000 * 60 * 60 * 24 * 7
export const MONTH_IN_MS = 18144000000; //1000 * 60 * 60 * 24 * 7 * 30
export const THREE_MONTHS_IN_MS = 54432000000; //1000 * 60 * 60 * 24 * 7 * 30 * 3

export const ITEMS_PER_PAGE = 10;
export const MAX_PAGINATION_ITEMS_ONE_DIRECTION_TO_SHOW = 4;
export const ITEMS_PER_PAGE_OPTIONS = [10, 20, 30, 50];

export const HINT = {
  AUDIO: 'audio',
  CUSTOM_FIELD: 'custom-field',
  IMAGE: 'image',
  NONE: 'none',
  TRANSLATION: 'translation',
  WORD_FIRST_LETTER: 'word-first-letter',
  WORD_LENGTH: 'word-length',
};

export const HINT_TYPES = [
  { value: HINT.AUDIO, text: 'Audio' },
  { value: HINT.CUSTOM_FIELD, text: 'Custom field' },
  { value: HINT.WORD_FIRST_LETTER, text: 'First letter' },
  { value: HINT.IMAGE, text: 'Image' },
  { value: HINT.NONE, text: 'None' },
  { value: HINT.TRANSLATION, text: 'Translation' },
  { value: HINT.WORD_LENGTH, text: 'Word length' },
];

export const DEFAULT_SORTING_OPTIONS = [
  { title: 'Date added', sortingFuncName: 'date-added' },
  { title: 'Date rated', sortingFuncName: 'date-rated' },
  { title: 'Date viewed', sortingFuncName: 'date-viewed' },
  { title: 'Random', sortingFuncName: 'random' },
  { title: 'Rating', sortingFuncName: 'rating' },
  { title: 'Word', sortingFuncName: 'id' },
];

export const QUESTION = {
  AUDIO: 'audio',
  CUSTOM_FIELD: 'custom-field',
  IMAGE: 'image',
  PHRASE: 'phrase',
  TRANSLATION: 'translation',
  WORD: 'word',
};

export const QUESTION_TYPES = [
  { value: QUESTION.AUDIO, text: 'Audio' },
  { value: QUESTION.CUSTOM_FIELD, text: 'Custom field' },
  { value: QUESTION.IMAGE, text: 'Image' },
  { value: QUESTION.PHRASE, text: 'Phrase' },
  { value: QUESTION.TRANSLATION, text: 'Translation' },
  { value: QUESTION.WORD, text: 'Word' },
];

export const RATE_BY = {
  SELF_RATING: 'self-rating',
  TYPE_IN: 'type-in',
};
