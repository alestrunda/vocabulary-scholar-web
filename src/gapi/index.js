/*global window*/

//create empty file, only set file meta
export const gapiCreateAppFile = fileName =>
  window.gapi.client.drive.files.create({
    name: fileName,
    parents: ['appDataFolder'],
    mimeType: 'application/json',
  });

//get the app file, returns promise
export const gapiGetAppFileID = fileName =>
  window.gapi.client.drive.files
    .list({
      spaces: 'appDataFolder',
      pageSize: 1, //app uses only one file
      fields: 'files(id, name)', //specify which fields should api return
    })
    .then(response => {
      const appStateFileIDs = response.result.files
        .filter(file => file.name === fileName)
        .map(file => file.id);
      return appStateFileIDs[0];
    });

//load file content
export const gapiGetFileContentByID = fileID =>
  window.gapi.client.drive.files.get({
    fileId: fileID,
    alt: 'media',
  });

export const gapiSignIn = () => window.gapi.auth2.getAuthInstance().signIn();

export const gapiSignOut = () => window.gapi.auth2.getAuthInstance().signOut();

//update file content
export const gapiUpdateFile = (fileID, fileContentJson) =>
  window.gapi.client.request({
    path: `/upload/drive/v3/files/${fileID}`,
    method: 'PATCH',
    params: { uploadType: 'media' },
    body: fileContentJson,
  });
