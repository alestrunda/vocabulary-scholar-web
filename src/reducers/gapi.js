import {
  GAPI_CLIENT_STATUS_UPDATE,
  GAPI_CONNECTED,
  GAPI_SET_STORE_FILE_ID,
  GAPI_STORE_LOADED_FROM_DRIVE,
} from '../actions/gapi';

const gapi = (state = {}, action) => {
  let isStoreLoadedFromDrive;
  const { payload, type } = action;
  switch (type) {
    case GAPI_CLIENT_STATUS_UPDATE:
      isStoreLoadedFromDrive = payload.isClientSignedIn //reset to false if client signed out
        ? state.isStoreLoadedFromDrive
        : false;
      return {
        ...state,
        isClientSignedIn: payload.isClientSignedIn,
        isStoreLoadedFromDrive,
      };
    case GAPI_CONNECTED:
      return {
        ...state,
        isConnected: true,
      };
    case GAPI_SET_STORE_FILE_ID:
      return {
        ...state,
        storeFileID: payload.storeFileID,
      };
    case GAPI_STORE_LOADED_FROM_DRIVE:
      return {
        ...state,
        isStoreLoadedFromDrive: payload.isStoreLoadedFromDrive,
        storeFileID: payload.storeFileID,
      };
    default:
      return state;
  }
};

export default gapi;
