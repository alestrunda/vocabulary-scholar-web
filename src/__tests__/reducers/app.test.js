import appReducer from '../../reducers/app';
import { ITEMS_PER_PAGE } from '../../constants';
import {
  appInfoMessage,
  appLoadState,
  appSetSelectedPhrase,
  appStateNotLoaded,
  appStorageType,
  appWarning,
  appWarningClose,
  appWordListSetPage,
} from '../../actions/app';
import { gapiClientStatusUpdate } from '../../actions/gapi';

it('correct app initial state', () => {
  const action = { type: '' }; //passing no action should return initial state
  const initialState = {
    language: '',
    infoMessage: '',
    itemsPerPage: ITEMS_PER_PAGE,
    phraseSelected: '',
    wordListsActivePages: {},
  };
  const defaultState = undefined;

  expect(appReducer(defaultState, action)).toEqual(initialState);
});

it('does not modify state if action not recognized', () => {
  const action = { type: 'unknown', payload: { someProp: 'some value' } };
  const defaultState = {
    testProp: 'test value',
  };

  expect(appReducer(defaultState, action)).toEqual(defaultState);
});

it('handles APP_INFO_MESSAGE', () => {
  const message = 'test message';
  const action = appInfoMessage(message);
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('handles APP_LOAD_STATE', () => {
  const stateToLoad = {
    app: {},
  };
  const storageType = 'local-storage';
  const action = appLoadState(stateToLoad, storageType);
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('handles APP_SET_SELECTED_PHRASE', () => {
  const phrase = 'Lorem ipsum';
  const action = appSetSelectedPhrase(phrase);
  const expectedState = {
    phraseSelected: phrase,
  };
  const defaultState = {};

  expect(appReducer(defaultState, action)).toEqual(expectedState);
});

it('handles APP_STATE_NOT_LOADED', () => {
  const action = appStateNotLoaded();
  const expectedState = {
    isStateLoaded: false,
  };
  const defaultState = {};

  expect(appReducer(defaultState, action)).toEqual(expectedState);
});

it('handles APP_STORAGE_TYPE', () => {
  const action = appStorageType('local-storage');
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('handles APP_WARNING', () => {
  const action = appWarning('something went wrong', 'user-error');
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('handles APP_WARNING_CLOSE', () => {
  const action = appWarningClose('user-error');
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('sets active page on APP_WORD_LIST_SET_PAGE', () => {
  const action = appWordListSetPage(5);
  const defaultState = {
    wordListsActivePages: {},
  };

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});

it('updates active page on APP_WORD_LIST_SET_PAGE', () => {
  const listID = 'list1';
  const newPageNumber = 10;
  const action = appWordListSetPage(newPageNumber, listID);
  const defaultState = {
    wordListsActivePages: {},
  };
  defaultState.wordListsActivePages[listID] = 5;

  expect(appReducer(defaultState, action).wordListsActivePages[listID]).toBe(
    newPageNumber
  );
});

it('handles GAPI_CLIENT_STATUS_UPDATE', () => {
  const isClientSignedIn = true,
    isInitial = true;
  const action = gapiClientStatusUpdate(isClientSignedIn, isInitial);
  const defaultState = {};

  expect(appReducer(defaultState, action)).toMatchSnapshot();
});
