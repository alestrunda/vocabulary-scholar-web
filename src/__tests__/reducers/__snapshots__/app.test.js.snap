// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles APP_INFO_MESSAGE 1`] = `
Object {
  "infoMessage": "test message",
}
`;

exports[`handles APP_LOAD_STATE 1`] = `
Object {
  "isStateLoaded": true,
  "storageType": "local-storage",
}
`;

exports[`handles APP_STORAGE_TYPE 1`] = `
Object {
  "storageType": "local-storage",
}
`;

exports[`handles APP_WARNING 1`] = `
Object {
  "warning": "something went wrong",
  "warningType": "user-error",
}
`;

exports[`handles APP_WARNING_CLOSE 1`] = `
Object {
  "user-error": false,
}
`;

exports[`handles GAPI_CLIENT_STATUS_UPDATE 1`] = `Object {}`;

exports[`sets active page on APP_WORD_LIST_SET_PAGE 1`] = `
Object {
  "wordListsActivePages": Object {
    "undefined": 5,
  },
}
`;
